File: src/bug_resolution_radar/theme/semantic_colors.py
"""Semantic status/priority colors shared across backend and UI."""

from __future__ import annotations

import re
from typing import Dict, Iterable, Optional

from bug_resolution_radar.theme.design_tokens import (
    BBVA_GOAL_ACCENT_7,
    BBVA_GOAL_SURFACE_8,
    BBVA_NEUTRAL_SOFT,
    BBVA_SIGNAL_GREEN_1,
    BBVA_SIGNAL_GREEN_2,
    BBVA_SIGNAL_GREEN_3,
    BBVA_SIGNAL_ORANGE_2,
    BBVA_SIGNAL_RED_1,
    BBVA_SIGNAL_RED_2,
    BBVA_SIGNAL_RED_3,
    BBVA_SIGNAL_YELLOW_1,
)


def normalize_semantic_token(value: Optional[str]) -> str:
    txt = str(value or "").strip().lower()
    txt = txt.replace("_", " ").replace("-", " ")
    txt = re.sub(r"\s+", " ", txt).strip()
    return txt


STATUS_COLOR_BY_KEY: Dict[str, str] = {
    "new": BBVA_SIGNAL_RED_3,
    "ready": BBVA_SIGNAL_RED_3,
    "analysing": BBVA_SIGNAL_RED_3,
    "blocked": BBVA_SIGNAL_RED_3,
    "en progreso": BBVA_SIGNAL_ORANGE_2,
    "in progress": BBVA_SIGNAL_ORANGE_2,
    "to rework": BBVA_SIGNAL_ORANGE_2,
    "rework": BBVA_SIGNAL_ORANGE_2,
    "test": BBVA_SIGNAL_ORANGE_2,
    "ready to verify": BBVA_SIGNAL_ORANGE_2,
    "accepted": BBVA_SIGNAL_GREEN_3,
    "ready to deploy": BBVA_SIGNAL_GREEN_3,
    "deployed": BBVA_GOAL_ACCENT_7,
    "closed": BBVA_SIGNAL_GREEN_1,
    "resolved": BBVA_SIGNAL_GREEN_1,
    "done": BBVA_SIGNAL_GREEN_1,
    "open": BBVA_SIGNAL_YELLOW_1,
    "created": BBVA_SIGNAL_RED_3,
}

PRIORITY_COLOR_BY_KEY: Dict[str, str] = {
    "supone un impedimento": BBVA_SIGNAL_RED_1,
    "highest": BBVA_SIGNAL_RED_1,
    "high": BBVA_SIGNAL_RED_2,
    "medium": BBVA_SIGNAL_ORANGE_2,
    "low": BBVA_SIGNAL_GREEN_2,
    "lowest": BBVA_SIGNAL_GREEN_1,
}


def status_color(status: Optional[str]) -> str:
    return STATUS_COLOR_BY_KEY.get(normalize_semantic_token(status), BBVA_NEUTRAL_SOFT)


def priority_color(priority: Optional[str]) -> str:
    return PRIORITY_COLOR_BY_KEY.get(normalize_semantic_token(priority), BBVA_NEUTRAL_SOFT)


def status_color_map(statuses: Optional[Iterable[str]] = None) -> Dict[str, str]:
    if statuses is None:
        return {}
    return {str(status): status_color(str(status)) for status in statuses}


def flow_signal_color_map() -> Dict[str, str]:
    return {
        "created": BBVA_SIGNAL_RED_3,
        "closed": BBVA_SIGNAL_GREEN_2,
        "resolved": BBVA_SIGNAL_GREEN_2,
        "deployed": BBVA_GOAL_ACCENT_7,
        "open": BBVA_SIGNAL_YELLOW_1,
        "open_backlog_proxy": BBVA_SIGNAL_YELLOW_1,
    }


def priority_color_map() -> Dict[str, str]:
    return {
        "Supone un impedimento": BBVA_SIGNAL_RED_1,
        "Highest": BBVA_SIGNAL_RED_1,
        "High": BBVA_SIGNAL_RED_2,
        "Medium": BBVA_SIGNAL_ORANGE_2,
        "Low": BBVA_SIGNAL_GREEN_2,
        "Lowest": BBVA_SIGNAL_GREEN_1,
        "(sin priority)": BBVA_NEUTRAL_SOFT,
        "": BBVA_NEUTRAL_SOFT,
    }


def semantic_color_contract() -> Dict[str, object]:
    """Return the frontend-consumable semantic token contract."""
    return {
        "statusByKey": dict(STATUS_COLOR_BY_KEY),
        "priorityByKey": dict(PRIORITY_COLOR_BY_KEY),
        "neutral": BBVA_NEUTRAL_SOFT,
        "goalAccent": BBVA_GOAL_ACCENT_7,
        "goalSurface": BBVA_GOAL_SURFACE_8,
    }

File: frontend/src/lib/semanticColors.ts
import type { CSSProperties } from "react";

type SemanticConfig = {
  statusByKey: Record<string, string>;
  priorityByKey: Record<string, string>;
  neutral: string;
  goalAccent: string;
  goalSurface: string;
};

const SEMANTIC_CONFIG: SemanticConfig = {
  statusByKey: {
    new: "var(--bbva-status-intake)",
    ready: "var(--bbva-status-intake)",
    analysing: "var(--bbva-status-intake)",
    blocked: "var(--bbva-status-intake)",
    "en progreso": "var(--bbva-status-progress)",
    "in progress": "var(--bbva-status-progress)",
    "to rework": "var(--bbva-status-progress)",
    rework: "var(--bbva-status-progress)",
    test: "var(--bbva-status-progress)",
    "ready to verify": "var(--bbva-status-progress)",
    accepted: "var(--bbva-status-accepted)",
    "ready to deploy": "var(--bbva-status-accepted)",
    deployed: "var(--bbva-status-deployed)",
    closed: "var(--bbva-priority-lowest)",
    resolved: "var(--bbva-priority-lowest)",
    done: "var(--bbva-priority-lowest)",
    open: "var(--bbva-status-open)",
    created: "var(--bbva-status-intake)"
  },
  priorityByKey: {
    "supone un impedimento": "var(--bbva-priority-highest)",
    highest: "var(--bbva-priority-highest)",
    high: "var(--bbva-priority-high)",
    medium: "var(--bbva-priority-medium)",
    low: "var(--bbva-priority-low)",
    lowest: "var(--bbva-priority-lowest)"
  },
  neutral: "var(--bbva-neutral)",
  goalAccent: "var(--bbva-goal-accent)",
  goalSurface: "var(--bbva-goal-surface)"
};

function normalizeSemanticToken(value: string) {
  return String(value || "")
    .trim()
    .toLowerCase()
    .replaceAll("_", " ")
    .replaceAll("-", " ")
    .replace(/\s+/g, " ");
}

function colorWithAlpha(color: string, alpha: number) {
  const source = String(color || "").trim() || SEMANTIC_CONFIG.neutral;
  const token = source.replace(/^#/, "");
  const boundedAlpha = Math.min(1, Math.max(0, alpha));
  if (/^[0-9a-fA-F]{6}$/.test(token)) {
    const r = Number.parseInt(token.slice(0, 2), 16);
    const g = Number.parseInt(token.slice(2, 4), 16);
    const b = Number.parseInt(token.slice(4, 6), 16);
    return `rgba(${r}, ${g}, ${b}, ${boundedAlpha})`;
  }
  return `color-mix(in srgb, ${source} ${boundedAlpha * 100}%, transparent)`;
}

function chipPalette(color: string) {
  const normalized = String(color || "").trim().toUpperCase();
  const goalAccent = SEMANTIC_CONFIG.goalAccent.toUpperCase();
  if (normalized === goalAccent) {
    return {
      color: SEMANTIC_CONFIG.goalAccent,
      borderColor: colorWithAlpha(SEMANTIC_CONFIG.goalAccent, 0.64),
      backgroundColor: SEMANTIC_CONFIG.goalSurface
    };
  }
  return {
    color,
    borderColor: colorWithAlpha(color, 0.62),
    backgroundColor: colorWithAlpha(color, 0.16)
  };
}

export function statusColor(status: string) {
  return (
    SEMANTIC_CONFIG.statusByKey[normalizeSemanticToken(status)] ?? SEMANTIC_CONFIG.neutral
  );
}

export function priorityColor(priority: string) {
  return (
    SEMANTIC_CONFIG.priorityByKey[normalizeSemanticToken(priority)] ?? SEMANTIC_CONFIG.neutral
  );
}

export function semanticChipStyle(
  value: string,
  kind: "status" | "priority"
): CSSProperties {
  const color = kind === "priority" ? priorityColor(value) : statusColor(value);
  const palette = chipPalette(color);
  return {
    color: palette.color,
    borderColor: palette.borderColor,
    backgroundColor: palette.backgroundColor
  };
}

export function neutralChipStyle(fontSize = "0.8rem"): CSSProperties {
  return {
    color: "var(--bbva-text-muted)",
    borderColor: "var(--bbva-border-strong)",
    background:
      "color-mix(in srgb, var(--bbva-surface) 86%, var(--bbva-surface-2))",
    fontSize
  };
}

export function kanbanHeaderStyle(status: string, active: boolean): CSSProperties {
  const color = statusColor(status);
  return {
    color,
    borderColor: colorWithAlpha(color, active ? 0.72 : 0.45),
    backgroundColor: colorWithAlpha(color, active ? 0.2 : 0.12),
    boxShadow: active ? `0 0 0 3px ${colorWithAlpha(color, 0.18)}` : "none"
  };
}

export function semanticButtonStyle(
  value: string,
  kind: "status" | "priority",
  active: boolean
): CSSProperties {
  const color = kind === "priority" ? priorityColor(value) : statusColor(value);
  return {
    color,
    borderColor: colorWithAlpha(color, active ? 0.72 : 0.45),
    backgroundColor: colorWithAlpha(color, active ? 0.2 : 0.12),
    boxShadow: active ? `0 0 0 3px ${colorWithAlpha(color, 0.16)}` : "none"
  };
}
